airbyte-webapp/src/components/connectorBuilder/StreamTestingPanel/SchemaDiffView.tsx

```diff
@@ -0,0 +1,168 @@
+import { faWarning } from "@fortawesome/free-solid-svg-icons";
+import classNames from "classnames";
+import { diffJson, Change } from "diff";
+import { useField } from "formik";
+import merge from "lodash/merge";
+import React, { useMemo, useState } from "react";
+import { FormattedMessage } from "react-intl";
+import { useDebounce } from "react-use";
+
+import { Button } from "components/ui/Button";
+import { FlexContainer, FlexItem } from "components/ui/Flex";
+import { InfoBox } from "components/ui/InfoBox";
+
+import { StreamReadInferredSchema } from "core/request/ConnectorBuilderClient";
+import {
+  useConnectorBuilderFormState,
+  useConnectorBuilderTestState,
+} from "services/connectorBuilder/ConnectorBuilderStateService";
+
+import { formatJson } from "../utils";
+import styles from "./SchemaDiffView.module.scss";
+
+interface SchemaDiffViewProps {
+  inferredSchema: StreamReadInferredSchema;
+}
+
+interface Diff {
+  /**
+   * List of changes from current schema to detected schema
+   */
+  changes: Change[];
+  /**
+   * Formatted merged schema if merging in the detected schema changes the existing schema
+   */
+  mergedSchema?: string;
+  /**
+   * Flag if overriding the existing schema with the new schema would lose information
+   */
+  lossyOverride: boolean;
+}
+
+function getDiff(existingSchema: string | undefined, detectedSchema: object): Diff {
+  if (!existingSchema) {
+    return { changes: [], lossyOverride: false };
+  }
+  try {
+    const existingObject = existingSchema ? JSON.parse(existingSchema) : undefined;
+    const mergedSchemaPreferExisting = formatJson(merge({}, detectedSchema, existingObject));
+    const changes = diffJson(existingObject, detectedSchema);
+    // The override would be lossy if lines are removed in the diff
+    const lossyOverride = changes.some((change) => change.removed);
+    return {
+      changes,
+      mergedSchema: mergedSchemaPreferExisting !== existingSchema ? mergedSchemaPreferExisting : undefined,
+      lossyOverride,
+    };
+  } catch {
+    return { changes: [], lossyOverride: true };
+  }
+}
+
+export const SchemaDiffView: React.FC<SchemaDiffViewProps> = ({ inferredSchema }) => {
+  const { testStreamIndex } = useConnectorBuilderTestState();
+  const { editorView } = useConnectorBuilderFormState();
+  const [field, , helpers] = useField(`streams[${testStreamIndex}].schema`);
+  const formattedSchema = useMemo(() => inferredSchema && formatJson(inferredSchema), [inferredSchema]);
+
+  const [schemaDiff, setSchemaDiff] = useState<Diff>(() =>
+    editorView === "ui" ? getDiff(field.value, inferredSchema) : { changes: [], lossyOverride: false }
+  );
+
+  useDebounce(
+    () => {
+      if (editorView === "ui") {
+        setSchemaDiff(getDiff(field.value, inferredSchema));
```

Comment:
As part of the diffing the library is ordering all keys alphabetically: https://github.com/kpdecker/jsdiff/blob/master/src/diff/json.js#L72

This is a little annoying in this context, I'm not sure how bad it is though. An idea I have is to always order keys alphabetically in all the schemas to enforce some consistency as part of the `formatJson` helper. What do you think? This is what I implemented on the PR.

The only simple way I see to do the diffing respecting the original source order is to do a line-diff on the stringified schema instead of diffing the object directly, but that would cause other weird cases:
* Adding a new property will cause a difference because it adds a comma to the end of the previously last property:
```
"a": {
  "first: "abc"
}
```

to 
```
"a": {
  "first: "abc",
  "second": "def"
}
```

has the following diff:
```
 "a": {
-  "first: "abc"
+  "first: "abc",
+  "second": "def"
}
```

* As the diffing algorithm doesn't know about object nesting anymore, it's possible diffs will span multiple nested objects which gets confusing easily (like in git when adding and removing a function in the same place causes the zebra pattern of deleted, added and unchanged lines). Would be nice to avoid this.

Reply:
I agree that trying to do line-level diffs doesn't seem like a good idea due to the reasons you listed.

I think ordering keys alphabetically everywhere should be fine. The only other ideas I have here are:
- Re-order the schemaDiff so that the keys match the `inferredSchema` key order (though with this approach we wouldn't know where to place keys that exist in declared schema but don't exist in the inferred schema)
- Fork (🤮) the json diff library to remove the `sort()` call or open an issue to make that sorting configurable

But I think ordering the keys alphabetically is probably the simplest approach that guarantees the most consistency, so I'm fine with that approach. We can always revisit if people complain about the alphabetical sorting

Reply:
> Re-order the schemaDiff so that the keys match the inferredSchema key ord

I guess it's possible somehow, but the data structure returned from the library is line based, so ordering the keys based on these will get some pretty complex code parsing property names out of strings while tracking the nesting level. Definitely think it's worth pushing that out.